src/chain_rules.jl

```diff
@@ -4,27 +4,95 @@ ChainRulesCore.@scalar_rule +(x::APL) true
 ChainRulesCore.@scalar_rule -(x::APL) -1
 
 ChainRulesCore.@scalar_rule +(x::APL, y::APL) (true, true)
+function plusconstant1_pullback(Δ)
+    return ChainRulesCore.NoTangent(), Δ, coefficient(Δ, constantmonomial(Δ))
+end
+function ChainRulesCore.rrule(::typeof(plusconstant), p::APL, α)
+    return plusconstant(p, α), plusconstant1_pullback
+end
+function plusconstant2_pullback(Δ)
+    return ChainRulesCore.NoTangent(), coefficient(Δ, constantmonomial(Δ)), Δ
+end
+function ChainRulesCore.rrule(::typeof(plusconstant), α, p::APL)
+    return plusconstant(α, p), plusconstant2_pullback
+end
 ChainRulesCore.@scalar_rule -(x::APL, y::APL) (true, -1)
 
 function ChainRulesCore.frule((_, Δp, Δq), ::typeof(*), p::APL, q::APL)
     return p * q, MA.add_mul!!(p * Δq, q, Δp)
 end
+
+function _adjoint_mult(op::F, ts, p, Δ) where {F<:Function}
+    for t in terms(p)
+        c = coefficient(t)
+        m = monomial(t)
+        for δ in Δ
+            if divides(m, δ)
+                coef = op(c, coefficient(δ))
+                mono = _div(monomial(δ), m)
+                push!(ts, term(coef, mono))
+            end
+        end
+    end
+    return polynomial(ts)
+end
+function adjoint_mult_left(p, Δ)
+    ts = MA.promote_operation(*, MA.promote_operation(adjoint, termtype(p)), termtype(Δ))[]
+    return _adjoint_mult(ts, p, Δ) do c, d
+        c' * d
+    end
+end
+function adjoint_mult_right(p, Δ)
+    ts = MA.promote_operation(*, termtype(Δ), MA.promote_operation(adjoint, termtype(p)))[]
+    return _adjoint_mult(ts, p, Δ) do c, d
+        d * c'
+    end
+end
+
 function ChainRulesCore.rrule(::typeof(*), p::APL, q::APL)
     function times_pullback2(ΔΩ̇)
-        #ΔΩ = ChainRulesCore.unthunk(Ω̇)
-        #return (ChainRulesCore.NoTangent(), ChainRulesCore.ProjectTo(p)(ΔΩ * q'), ChainRulesCore.ProjectTo(q)(p' * ΔΩ))
-        return (ChainRulesCore.NoTangent(), ΔΩ̇ * q', p' * ΔΩ̇)
+        return (ChainRulesCore.NoTangent(), adjoint_mult_right(q, ΔΩ̇), adjoint_mult_left(p, ΔΩ̇))
```

Comment:
The previous definition was correct for the scalar product `dot(p, q) = p * q`, the current one is correct for the scalar product `dot(p, q) = dot(coefficients(p), coefficients(q))` (assuming they have the same monomials).

Reply:
This second scalar product corresponds to the `_dot` we use for testing `differentiate`. This means that for `differentiate` we assume the other scalar product